pdf/core/parser.go

```diff
@@ -1333,11 +1334,45 @@ func (parser *PdfParser) loadXrefs() (*PdfObjectDictionary, error) {
 // Return the closest object following offset from the xrefs table.
 func (parser *PdfParser) xrefNextObjectOffset(offset int64) int64 {
 	nextOffset := int64(0)
-	for _, xref := range parser.xrefs {
-		if xref.Offset > offset && (xref.Offset < nextOffset || nextOffset == 0) {
-			nextOffset = xref.Offset
+
+	if len(parser.xrefs.ObjectMap) == 0 {
+		return 0
+	}
+
+	if len(parser.xrefs.sortedObjects) == 0 {
+		count := 0
+		for _, xref := range parser.xrefs.ObjectMap {
+			if xref.Offset > 0 {
+				count++
+			}
+		}
+		if count == 0 {
+			// No objects with offset.
+			return 0
+		}
+		parser.xrefs.sortedObjects = make([]XrefObject, count)
+
+		i := 0
+		for _, xref := range parser.xrefs.ObjectMap {
+			if xref.Offset > 0 {
+				parser.xrefs.sortedObjects[i] = xref
+				i++
+			}
 		}
+
+		// Sort by offset, descending.
+		sort.Slice(parser.xrefs.sortedObjects, func(i, j int) bool {
+			return parser.xrefs.sortedObjects[i].Offset < parser.xrefs.sortedObjects[j].Offset
+		})
 	}
+
+	i := sort.Search(len(parser.xrefs.sortedObjects), func(i int) bool {
```

Comment:
This seems to be equivalent to:
```
	for _, obj := range parser.xrefs.sortedObjects {
		if obj.Offset >= offset {
			return obj.Offset
		}
	}
	return 0
```

Reply:
Yes, except sort.Search uses binary search which is very efficient for sorted data.